File: 2_backend/src/app/factcheck/parser.py
"""Contains parsing functions for handling fact check api response."""

from __future__ import annotations

import json

from app.factcheck.models import (
    AlignmentResult,
    Claim,
    FactCheck,
    SourceAlignment,
)


def parse_claims(response: str) -> list[Claim]:
    """Parse the list of claims from the LLM response.

    Handles two formats:
      - With brackets:  [claim_1 | claim_2 | claim_N]
      - Without brackets: claim_1 | claim_2 | claim_N
    """
    text = response.strip()
    if "[" in text and "]" in text:
        start = text.index("[")
        end = text.index("]")
        claims_str = text[start + 1 : end]
    else:
        claims_str = text
    return [Claim(content=claim.strip()) for claim in claims_str.split("|")]


def parse_alignment_result(
    response: str, claim: Claim, fact_checks: list[FactCheck]
) -> AlignmentResult:
    """Parse the LLM response into an AlignmentResult."""
    start = response.index("[")
    end = response.rindex("]") + 1
    verdicts = json.loads(response[start:end])

    alignments = [
        SourceAlignment(
            source=fact_checks[v["id"]].source,
            verdict=v["verdict"],
        )
        for v in verdicts
        if v["id"] < len(fact_checks) and v["relevant"]
    ]

    return AlignmentResult(claim=claim.content, alignments=alignments)

File: 2_browser-extension/src/pipeline/parser.ts
import { AlignmentResult, Claim, FactCheck, SourceAlignment } from "./types";

export function parseClaims(response: string): Claim[] {
  let claimsStr = response.trim();

  if (claimsStr.includes("[") && claimsStr.includes("]")) {
    const start = claimsStr.indexOf("[");
    const end = claimsStr.lastIndexOf("]");
    claimsStr = claimsStr.slice(start + 1, end);
  }

  return claimsStr
    .split("|")
    .map((claim) => ({ content: claim.trim() }))
    .filter((claim) => claim.content.length > 0);
}

export function parseAlignmentResult(
  response: string,
  claim: Claim,
  factChecks: FactCheck[],
): AlignmentResult {
  const start = response.indexOf("[");
  const end = response.lastIndexOf("]") + 1;

  let verdicts: {
    id: number;
    verdict: "CONTRADICTED" | "MISLEADING" | "UNVERIFIED";
    relevant: boolean;
  }[];

  try {
    if (start < 0 || end <= start) throw new Error("no JSON array in response");
    verdicts = JSON.parse(response.slice(start, end));
  } catch {
    // Malformed LLM output → treat as "no usable alignment" instead of crashing.
    return { claim: claim.content, alignments: [] };
  }

  const alignments: SourceAlignment[] = verdicts
    .filter((v) => v.id < factChecks.length && v.relevant)
    .map((v) => ({
      source: factChecks[v.id].source,
      verdict: v.verdict,
    }));

  return { claim: claim.content, alignments };
}
